### ISSUE TITLE
"Content-Disposition"-based file renaming (Xposted from sf.net)

### ISSUE THREAD

**Issue by lucisandor** [dvehrs/podget#2]
Since 2012 (v 1.11), Wget allows renaming files according to server's instructions, as given in the HTTP header "Content-Disposition". All it takes is adding the flag "--content-disposition" to wget's invocation.
With that flag the BBC podcast http://open.live.bbc.co.uk/mediaselector/5/redir/version/2.0/mediaset/audio-nondrm-download/proto/http/vpid/p033zrky.mp3 is renamed as Documentaries-20151001-SpainsBattleForTheBull.mp3.
I suggest we get "--content-disposition" activated in Podget.

**Comment by dvehrs** [comment]
OK, first let me say I haven't been ignoring this but rather trying to get my brain around both its potential and any implementation issues.

On the good side, I can understand why this feature would be desirable from both content provider and consumer perspectives.

The down side is not so rosy.  

First, the --content-disposition flag is still labeled as experimental in WGET's man page.  I'm not sure if that's because the manual page has not been updated yet or if the feature is still considered experimental.  As a general rule, I'm against using experimental features as they can't be counted on to remain the same for years.  

Second there is a potential implementation issue, use of the flag can result in extra round trip connections to the server for a "HEAD" request.  I'm not sure how to integrate that with Podget's existing file naming and directory functions.    

Third (also an implementation issue), can we enable this flag by default for all feeds?  Will it cause issues for feeds that do not use it?  Can we add this by merely enabling it as a WGET flag in podgetrc?

Now that's how far I've made it so far.  I'm not necessarily for or against adding this feature.  I just don't have enough information yet.  It would help me do some initial testing if you could share the feed URL that you got the above MP3 URL from.  Is this a private feed or is it public?

Happy New Year.

Dave

**Comment by lucisandor** [comment]
Happy New Year to you too, and many thanks for your continuing support.
The BBC feeds are public.
I understand your worries. I was concerned that podget may be unable to capture the new file name, thus disabling the various ```FILENAME_FORMATFIX``` renaming strategies. I should try it for a bit. I will duplicate /usr/bin/podget, make the ```Content-disposition``` modification on the line calling ```wget```, and let the copy run in parallel with the original for a while.

**Comment by lucisandor** [comment]
So I have tried adding ```--content-disposition``` to wget options, but it won't work, because the final file name is computed in advance, from the URL.

A workaround would be to insert something like:

```
                        # This fix enables file renaming based on server sent Content-disposition.
                        if [ ! -z "${FILENAME_FORMATFIX11+set}" ] && [ "${FILENAME_FORMATFIX11}" -gt 0 ] ; then
						    URL_FILENAME=$(wget --server-response -q -O - "${URL}" 2>&1 | grep "Content-Disposition:" | tail -1 | awk 'match($0, /filename=(.+)/, f){ print f[1] }' | tr -d "\"'" )
                            if [ ${VERBOSITY} -ge 3 ] ; then
                                echo "FILENAME FORMAT(11) FIXED: ${URL_FILENAME}"
                            fi
                        fi
```

This fits best immediately after ```URL_BASE="${URL%/*}"``` in the main procedure (line 1781). My code gets the file name using Content-disposition, and makes it available to all other FORMATFIX-es, but allows actual download proceed later, using regular wget settings.

Line 462 would require something along the lines of:

```
# Filename Cleanup 12:
#
# This is an attempt to fix feeds where enclosures are renamed by the served through the 
# Content-disposition mechanism.
#
# 0 == disabled
# 1 == enabled (default)
FILENAME_FORMATFIX11=1
```

**Comment by lucisandor** [comment]
Actually, the file name could be within quotes or not. Please use this:

```
# This fix enables file renaming based on server sent Content-disposition.
if [ ! -z "${FILENAME_FORMATFIX11+set}" ] && [ "${FILENAME_FORMATFIX11}" -gt 0 ] ; then
	CONTENT_FILENAME=$(wget --server-response -q -O - "${URL}" 2>&1 | grep "Content-Disposition:" | tail -1 | sed s/.*filename=// )
	if [ ! -z "$CONTENT_FILENAME" ] ; then
		if [[ $CONTENT_FILENAME =~ ^\" ]] ; then
			CONTENT_FILENAME=$(echo "${CONTENT_FILENAME}" | sed 's/"\(.*\)".*/\1/' )
		elif [[ $CONTENT_FILENAME =~ ^\' ]] ; then
			CONTENT_FILENAME=$(echo "${CONTENT_FILENAME}" | sed "s/'\(.*\)'.*/\1/" )
		else
			 CONTENT_FILENAME=$(echo "${CONTENT_FILENAME}" | sed 's/^\(\S\+\).*/\1/' )
		fi
	fi
	if [ ! -z "$CONTENT_FILENAME" ] ; then
		URL_FILENAME=$CONTENT_FILENAME
	fi
	if [ ${VERBOSITY} -ge 3 ] ; then
		echo "FILENAME FORMAT(11) FIXED: ${URL_FILENAME}"
	fi
fi
```

**Comment by dvehrs** [comment]
OK, interesting method but one small problem.  This fix only appears to be a viable solution for downloading small files or fast connections because in its attempt to get the newly assigned file name, it actually downloads the whole file.  So this FILENAME_FORMATFIX11 would effectively double the time to download each item on the feed.  

I tested the CONTENT_FILENAME step with one enclosure URL from a BBC  feed and to get the file name, I downloaded & discarded 12.4MB of data.  Now on my connection that only took 8.3 seconds but I have feeds in my collection that commonly have files over 100MB, so even on my fast connection this would impose an unacceptable delay.   

For a initial attempt you've proven that the concept is viable, however we have room for improvement.   What I think we need to find, is a way to get the value of the Content-Disposition field without having to download the entire file.  I'm not sure if that can be done with WGET but it merits looking into.  

Feed I tested with: 
http://www.bbc.co.uk/programmes/p02nq0lx/episodes/downloads.rss

File I downloaded to get name:  
http://open.live.bbc.co.uk/mediaselector/5/redir/version/2.0/mediaset/audio-nondrm-download-low/proto/http/vpid/p035nwbf.mp3

Content-Disposition: attachment; filename="Documentaries-20151020-MoreThanOneKindOfLove.mp3"

**Comment by lucisandor** [comment]
I was disappointed by the double download too. I think there is no way with wget alone. There is a ```--spider``` option, but it doesn't follow the redirections that lead to the BBC enclosures. Annoying, but that's BBC's and CDN's choice.

Also, above, the ```grep "Content-Disposition:"``` should be ```grep "Content-Disposition:" | grep filename```.

**Comment by dvehrs** [comment]
OK, I've stumbled on something that may point us in the correct direction and it appears we're on the same path......

If we use 'wget --server-response --spider' then it downloads just the headers and not the whole file.  However there is one problem, the file URL I used above points at a redirect and not an actual file on the server.  Unfortunately when I use the '--spider' option, it does not tell you the redirect URL.  Rather it merely tells you "Remote file does not exist -- broken link!!!".  

But if I use the redirect URL (in the 'Location:' header) from a full download of the file then I can use that URL with '--spider' to get the Content-Disposition field without downloading the file.   So the question is how to get a wget spider to follow redirects?

FILE REDIRECT with Content-Disposition:
http://aod-pod-ww-live.edgesuite.net/mpg_mp3_vlow/2c978993505b148f01508121f5286510--audio--thedocumentary-tue--201015-wk42-morethanonekindoflove1of1--pod-_mpg_mp3_vlow.mp3?__gda__=1452124828_b8e91162f55c64831c804d45d7061261

**Comment by dvehrs** [comment]
https://unix.stackexchange.com/questions/89282/how-to-figure-out-where-a-link-gets-redirected

It appears that WGET is not the ideal tool for this task.  I'm encountering more that a few recommendations to use CURL for it.  Not sure I'd like to expand Podget's dependencies but it may be necessary....

**Comment by lucisandor** [comment]
I tried what others recommend for cURL at https://stackoverflow.com/questions/34332845/get-content-disposition-filename-without-downloading . (Namely, issue a HEAD request, instead of GET)
It is not working, because the CDN is set to respond to HEAD with "403 Forbidden", regardless of the tool we use client-side. We need to GET.
I couldn't find a way to stop GETs at the first few bytes. That would be a way of saving bandwidth.

**Comment by dvehrs** [comment]
Everything we eliminate gets us one step closer to the answer.  We just need to find it.

**Comment by dvehrs** [comment]
OK, tentative good news.  I'm working on a possible solution to this problem.  By moving the renaming of the file to after it's been downloaded and capturing the --server-response output in a temporary file, I've been able to get the new file name in the same session as the download.  So far the only negative impact of this change is that Podget no longer has download progress information while the file is downloading.  As this only adversely affects Podget when it's running with verbosity >2 and with --content-disposition set in WGET_BASEOPTS, I'm inclined to tolerate it.  

I still need to do a variety of testing to see if the original filename from the server is generally safe for downloading & saving to an ext2/3/4 filesystem.  I hope I have enough feeds in my test collection to get good impression of how it will perform.  

Wish me luck!

**Comment by dvehrs** [comment]
Also it appears I might have found another potential issue with the BBC hosted feeds.  If you  have '-c' enabled in WGET_BASEOPTS, then the BBC will respond to all attempts to download with the same old "403 Forbidden" that we've seen so many other times.  

Of course, the solution to this is to simply use multiple podgetrc and serverlist files.  One combination for BBC feeds and one for everybody else may be enough to allow us to enable features as we need.

**Comment by lucisandor** [comment]
Yes, I removed the  "-c" from WGET_OPTIONS a while ago.
I'd be surprised to hear that the BBC is the only podcast provider that uses Akamai / Edgesuite to balance between locations.
To update you, I currently have:
```
CONTENT_FILENAME=$(wget --server-response -q -O /tmp/gunk "${URL}" 2>&1 )
if [[ $CONTENT_FILENAME =~ .*Content-Disposition.* ]] ; then
	CONTENT_FILENAME=$(echo "${CONTENT_FILENAME}" | grep "Content-Disposition:" )
	if [[ $CONTENT_FILENAME =~ .*filename.* ]] ; then
		CONTENT_FILENAME=$(echo "${CONTENT_FILENAME}" | grep "filename" | tail -1 | sed s/.*filename=// )
		if [[ $CONTENT_FILENAME =~ ^\" ]] ; then
			CONTENT_FILENAME=$(echo "${CONTENT_FILENAME}" | sed 's/\s*"\(.*\)".*/\1/' )
		elif [[ $CONTENT_FILENAME =~ ^\' ]] ; then
			CONTENT_FILENAME=$(echo "${CONTENT_FILENAME}" | sed "s/\s*'\(.*\)'.*/\1/" )
		else
			CONTENT_FILENAME=$(echo "${CONTENT_FILENAME}" | sed 's/\s*\(\S\S*\) .*/\1/' )
		fi
	fi
	else
	CONTENT_FILENAME=""
fi
if [ ! -z "$CONTENT_FILENAME" ] ; then
	URL_FILENAME=$CONTENT_FILENAME
fi
```
Even this does not cover the other funny things I heard people are doing with ```Content-disposition```.

**Comment by dvehrs** [comment]
OK speaking of funny things with Content-Disposition...

I've encountered a small problem with some files not being renamed.  I restarted the download and enabled a new "debug" mode so it does not delete the temporary files after attempting to rename the files.

The problem I found is the second of the files for these two Content-Disposition lines:

/tmp/user/1000/podget/p037zs8k.mp3.servresp:  Content-Disposition: attachment; filename="Documentaries-20151115-HomeBangladesh.mp3"
/tmp/user/1000/podget/p038bpmb.mp3.servresp:  Content-Disposition: attachment; filename="Documentaries-20151115-HomeBangladesh.mp3"

Two different files with the SAME filename= value!

And then I have four files whose server-response data has NO Content-Disposition field!

That's all I have so far.  I'm going to let this run go for a while longer and see what the total issues are at the end.

**Comment by dvehrs** [comment]
OK after a full run on the BBC feed listed in previous messages.

300 items downloaded.
255 renamed.
44 did not have a Content-Disposition item in their server-reply.
1 had the same filename defined as an earlier downloaded item (the one in previous message).

So almost a 1//6 ratio for issues. 

On the plus side, the method I've hacked together works 5/6 of the time with only one download per item.  This might merit an additional warming label to alert people to potential problems if they do attempt to use the --content-disposition flag.  

Now I just have to continue testing it with feeds that do not use the enable support for the --content-disposition flag.

**Comment by dvehrs** [comment]
I need to learn how to upload the 'development' branch I'm using to test it.  Once I figure out how to do that, it can become part of the Github record.

**Comment by lucisandor** [comment]
While we are at it, I wrote some very similar code (i.e., downloading twice) to deal with another HTTP header, "Location".
For example http://feeds.nature.com/nature/podcast/current is listing all the enclosures as "media.mp3". The only way to make it manageable with current podget is to use FORMATFIX4. But when one tries downloading the first enclosure ( http://feeds.nature.com/~r/nature/podcast/current/~5/VlzQtiUu-0Q/media.mp3 ) with ```wget --content-dispositoon``` or Firefox, one is redirected to http://az592690.vo.msecnd.net/media/channels/nature/acasts/naturepodcast-7january2016/US/ALL/ALL/nature-naturepodcast-7january2016.mp3?ts=1452103327528 , which provides unique names and removes the need for FORMATFIX4.
I got:
```
elif [[ $HEADERS =~ .*Location.* ]] ; then
	HEADERS=$(echo "${HEADERS}" | grep "Location:" | tail -1 | sed s/.*Location:// )
	if [[ $HEADERS =~ ^\" ]] ; then
		URL_FILENAME=$(echo "${HEADERS}" | sed 's/\s*"\(.*\)".*/\1/' )
	elif [[ $HEADERS =~ ^\' ]] ; then
		URL_FILENAME=$(echo "${HEADERS}" | sed "s/\s*'\(.*\)'.*/\1/" )
	else
		URL_FILENAME=$(echo "${HEADERS}" | sed 's/\s*\(\S\S*\) .*/\1/' )
	fi
	URL_FILENAME="${URL_FILENAME##*/}"
fi
```
I got all these double-downloading fixes in my copy of podget. If you think it is too hard, feel free to not implement this feature.

**Comment by dvehrs** [comment]
OK, I'm currently toying with an idea that will allow you to enable --content-disposition on a per feed basis rather than globally for all feeds.  The basics appear to work but its not fully tested yet.  

I wonder if this isn't another feature that should be enabled for select feeds rather than globally.

I did look at the files from the nature feed and the first one does have the format you cite, however one from near the end of the list was formatted like so nature-17f02738-d998-4bf2-b547-09fa9297eb50.mp3.  Now that may be ugly but it also should be unique and remove the need for FORMATFIX4.  And the ugly filename is definitely better than media.mp3.

Side Note:  We should try to avoid mixing multiple issues into one thread.  Doing so makes it much more difficult to know when to close a thread without going back and reading it in its entirety.  It also makes it more difficult to divide the issues up into individual test branches for git.  Now in this case I probably would have merged it with this thread so no real harm but lets try to keep each thread to one issue or fix.

**Comment by dvehrs** [comment]
OK, I've got a basic framework for renaming files based on the value from the final Location tag in their server response.  I've got it and the content-disposition option enabled on a per podcast basis rather than globally.  Plus in the future I may have to enable more flags to allow some customization of individual feed options.

From the Nature feed, I downloaded 141 items, 59 of which were human readable.  The other, while cryptic, were at least unique for each.  

On to more testing.....

**Comment by dvehrs** [comment]
OK, the content-disposition part is basically working.  However I suspect I've run into a small issue with the feed provider for it.  It does not appear to consistently provide the same information for each file.  I've done a second download of the entire feed and there are differences from what was downloaded & renamed for the first session.  A total of 85 differences between the two sessions.  Files got renamed in one but not the other.  It appears to be having issues approximately on just over 40 items per session so that means it got it right about 260 times out of 300 in each session.  I may need to find another feed provider that uses content-disposition to see if the problem is with my implementation or is it with the feed.

**Comment by dvehrs** [comment]
On the plus side, the similar system I have for getting the filename from the Location tag has worked 100% for two sessions. 

That leads me to wonder what the cause is for the --content-disposition handling.

**Comment by lucisandor** [comment]
OK, here is another feed where the file is renamed by Content-disposition: http://feeds.bmj.com/bmj/podcasts . Its enclosures are redirected to Soundcloud. To be honest, in this case, I prefer the initial enclosure filename.
I mentioned ```Location:``` only because my modifications to podget process it at the same time with ```Content disposition:```.

**Comment by dvehrs** [comment]
OK, initial tests with this new feed renames 100% of the files.  Although I do agree with you that the initial filenames are better than those specified by the Content-Disposition tag.  So that appears to be one more piece of evidence that the problems I'm experiencing with the BBC feed are because of something happening on their end.

As for the Location: issue, it was fortuitous that you mentioned it here because like you, I was able to leverage what I'd already built to address it as well.  I may be just over-sensitive to people adding secondary issues to threads that make closing the issue more work.  And it's not like you're alone, while I've been working on this, I had another bug report submitted that was trivial to fix so I integrated it into this git branch.  Probably shouldn't have but I did.  So sue me.  ;-P

**Comment by dvehrs** [comment]
OK, I'm getting more happy with this.  I've run the BBC feed six times and differences compared with the last session range from 80 to 113.  I want to test it with a few more feeds from my test collection but so far it looks pretty good.

**Comment by dvehrs** [comment]
OK good news for the BBC feed issue.  I've created a second tag that can be added on a per feed basis in the serverlist.  This tag tells podget to not add the file to the completed log if it fails to get a name from the Content-Disposition tag.  You still have the file with its initial name downloaded so you can listen to it but for every future session it will reattempt to download that file to see if it can get the filename.  Over the course of 6 sessions I was able to reduce my unnamed files from about 15% to under 2%.  

The downside to using this tag is you either have to enable it from the first time you attempt to download a feed or when you do, you have to manually remove any file URLs you wish to retry from the COMPLETED log manually.

**Comment by dvehrs** [comment]
OK, I figured out how to upload the 'development' branch to Github (and it was easier than I anticipated).   So you can view my testing copy in the 'dev' branch.  It's only 4 commits ahead of the primary branch so if anything I should probably commit more often.  If you have the time to take a look, I'd appreciate your opinion.

**Comment by lucisandor** [comment]
I tried the ```dev``` script, allowing it to create its own configuration. It downloaded the first enclosure, and exited with 
```
UTF-16 Loop running.
No config file found, exiting loop.
```
I am not sure if that is the desired behavior.

I also fed it my regular podget configuration, to which I added ```OPT_CONTENT_DISPOSITION=1```. It didn't rename the enclosures based on HTTP headers. I have ```FILENAME_FORMATFIX4=0``` and ```WGET_BASEOPTS="-nH"```, if that matters.

**Comment by dvehrs** [comment]
OK, the UTF-16 message is normal handling for a non-existent UTF-16 serverlist file.

OK, OPT_CONTENT_DISPOSITION does not get added in the podgetrc file, nor does it need to be set =1.  Rather you simply add it to the feed line in your serverlist file that you wish to enable Content-Disposition handling for.

Example:
http://somesite.com/rss/feed  CATEGORY Feed Name OPT_CONTENT_DISPOSITION

On the other hand, if you wish to enable Content-Disposition handling on a global basis for all feeds, then you add '--content-dispostion' to WGET_BASEOPTS.

This is similar to the option for getting the filename from the final Location tag after redirects.  To enable that you add OPT_FILENAME_LOCATION to the feed line in your serverlist file.

Example:
http://somesite.com/rss/feed CATEGORY Feed Name OPT_FILENAME_LOCATION

The third new option is OPT_DISPOSITION_FAIL and it only works when Content-Disposition handling is enabled (either globally or for that feed).

Example:
http://somesite.com/rss/feed  CATEGORY Feed Name OPT_CONTENT_DISPOSITION OPT_DISPOSITION_FAIL

 [ This may be the best combination of options for the BBC feed that we've been testing against ]

One last note, enabling Content-Disposition handling will not go back and get names for items that you previously downloaded and are marked as done in your LOG_COMPLETE.  If you wish to have those renamed, then you need to remove their URLS from the log and run a fresh session.

**Comment by dvehrs** [comment]
OK, I uploaded version 0.7.10 to Github and to Sourceforge with these fixes.  So it's live now.

Let me know if you run into any other issues.

Thanks again!

Dave

**Comment by lucisandor** [comment]
I have been using the Git version for a while without problems. BBC feeds never looked smarter :)

**Comment by dvehrs** [comment]
Good to know.  I was able to test it with a few other feeds for a good result so I decided to include it in the release version this last week.  I had a couple of 'unbound variable' errors reported and fixed in 0.7.12.  I don't know if anyone else is using this feature yet but I did recommend it to one user that was having a similar problem.  I downloaded 15GB of files from the feeds that he was having trouble with and the content disposition handling appears to be exactly what he needed.  I'm not sure I would have even attempted to look for it before so that's another user that was help because of what you initially proposed.  So more kudos to you for the good idea!

Thanks again!

Dave
